symphony-bdk-core/src/main/java/com/symphony/bdk/core/client/BearerEnabledApiClient.java

```diff
@@ -0,0 +1,96 @@
+import com.symphony.bdk.http.api.ApiException;
+import com.symphony.bdk.http.api.ApiResponse;
+import com.symphony.bdk.http.api.Pair;
+import com.symphony.bdk.http.api.util.TypeReference;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import lombok.extern.slf4j.Slf4j;
+import org.apiguardian.api.API;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An {@link ApiClient} implementation used when common jwt feature is enabled.
+ * It contains an {@link ApiClient} that will allow the usage of the Authentication header in each API call.
+ * It will also remove the sessionToken header from the request.
+ * Before adding the jwt as Authorization header the client is checking the jwt expiration, if the jwt is expired
+ * re-authentication will be triggered and the jwt will be refreshed.
+ */
+
+@Slf4j
+@API(status = API.Status.EXPERIMENTAL)
+public class BearerEnabledApiClient implements ApiClient {
+  private final ApiClient apiClient;
+  private final AuthSession authSession;
+
+  public BearerEnabledApiClient(ApiClient apiClient, AuthSession authSession) {
+    this.apiClient = apiClient;
+    this.authSession = authSession;
+  }
+
+  @Override
+  public <T> ApiResponse<T> invokeAPI(String path, String method, List<Pair> queryParams, Object body,
+      Map<String, String> headerParams, Map<String, String> cookieParams, Map<String, Object> formParams, String accept,
+      String contentType, String[] authNames, TypeReference<T> returnType) throws ApiException {
+    String authorizationToken = authSession.getAuthorizationToken();
+    if (authorizationToken != null) {
+      refreshTokenIfNeeded(authorizationToken);
+      headerParams.remove("sessionToken");
+      headerParams.put("Authorization", authSession.getAuthorizationToken());
+    }
+    return apiClient.invokeAPI(path, method, queryParams, body, headerParams, cookieParams, formParams, accept,
+        contentType, authNames, returnType);
+  }
+
+  @Override
+  public String getBasePath() {
+    return apiClient.getBasePath();
+  }
+
+  @Override
+  public String parameterToString(Object param) {
+    return apiClient.parameterToString(param);
+  }
+
+  @Override
+  public List<Pair> parameterToPairs(String collectionFormat, String name, Object value) {
+    return apiClient.parameterToPairs(collectionFormat, name, value);
+  }
+
+  @Override
+  public String selectHeaderAccept(String... accepts) {
+    return apiClient.selectHeaderAccept(accepts);
+  }
+
+  @Override
+  public String selectHeaderContentType(String... contentTypes) {
+    return apiClient.selectHeaderContentType(contentTypes);
+  }
+
+  @Override
+  public String escapeString(String str) {
+    return apiClient.escapeString(str);
+  }
+
+  /**
+   * Trigger jwt refresh if expired
+   */
+  private void refreshTokenIfNeeded(String authorizationToken) {
+    try {
+      Long expirationTime = JwtHelper.extractExpirationDate(authorizationToken);
```

Comment:
maybe we should do it once only when we receive the token and store this info is the AuthSession?

Reply:
done here: 4ff1b009f7b2ba82e38d8a8dc17c8ad980e7b1e9